src/shared/libfido2-util.c

```diff
@@ -195,6 +195,122 @@ static int verify_features(
+         * 2: https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html#authenticatorGetAssertion (in step 5)
+         */
+        if (has_up)
+                r = sym_fido_assert_set_up(a, FIDO_OPT_FALSE);
+        else
+                r = sym_fido_assert_set_up(a, FIDO_OPT_OMIT);
+        if (r != FIDO_OK)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                       "Failed to set assertion user presence: %s", sym_fido_strerr(r));
+
+
+        /* According to CTAP 2.1 specification, if the credential is marked as userVerificationRequired,
+         * we may pass pinUvAuthParam parameter or set uv option to true in order to check whether the
+         * credential is in the token. Here we choose to use PIN (pinUvAuthParam) to achieve this.
+         * If we don't do that, the authenticator will remove the credential from the applicable
+         * credentials list, hence CTAP2_ERR_NO_CREDENTIALS error will be returned.
+         * Please see
+         * https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-getAssert-authnr-alg
+         * step 7.4 for detailed informatrion.
+         */
+        if (FLAGS_SET(flags, FIDO2ENROLL_UV) && has_uv) {
+                if (strv_isempty(pins))
+                        r = FIDO_ERR_PIN_REQUIRED;
+                else
+                        STRV_FOREACH(i, pins) {
+                                r = sym_fido_dev_get_assert(d, a, *i);
+                                if (r != FIDO_ERR_PIN_INVALID)
+                                        break;
+                        }
+
+        } else
+                r = sym_fido_dev_get_assert(d, a, NULL);
+
+        if (r != FIDO_OK && r != FIDO_ERR_NO_CREDENTIALS) {
+                switch (r) {
+                        case FIDO_ERR_PIN_REQUIRED:
+                                return log_error_errno(SYNTHETIC_ERRNO(ENOANO),
+                                                       "Security token requires PIN.");
+                        case FIDO_ERR_PIN_INVALID:
+                                return log_error_errno(SYNTHETIC_ERRNO(ENOLCK),
+                                                       "PIN of security token incorrect.");
+                        case FIDO_ERR_PIN_BLOCKED:
+                                return log_error_errno(SYNTHETIC_ERRNO(EOWNERDEAD),
+                                                       "PIN of security token is blocked, please remove/reinsert token.");
+                        default:
+                                return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                                       "Failed to get assertion result: %s", sym_fido_strerr(r));
+                }
+        }
+
+        if (r == FIDO_ERR_NO_CREDENTIALS)
+                *ret_in_token = false;
+        else
+                *ret_in_token = true;
+
+        return 0;
```

Comment:
commonly across our codebase we have functions that return "extended booleans", i.e. they return == 0 as "false", > 0 as "true" and < 0 as "failure" (with the precise code indicating an error code.

please follow this here too, i.e. fold the `ret_in_token` return parameter into the actual return value, by return it as 0 (if token doesn't have the creds) or 1 (if it has) or -EIO (or some other errno) if we couldn#t determine that.